src/vs/platform/extensions/electron-main/workerMainProcessExtensionHostStarter.ts

```diff
@@ -0,0 +1,160 @@
+import { FileAccess } from 'vs/base/common/network';
+import { ILogService } from 'vs/platform/log/common/log';
+import { Worker } from 'worker_threads';
+import { IWorker, IWorkerCallback, IWorkerFactory, SimpleWorkerClient } from 'vs/base/common/worker/simpleWorker';
+import { IExtensionHostStarterWorkerHost } from 'vs/platform/extensions/node/extensionHostStarterWorker';
+import { ILifecycleMainService } from 'vs/platform/lifecycle/electron-main/lifecycleMainService';
+import { ExtensionHostStarter } from 'vs/platform/extensions/node/extensionHostStarter';
+
+class NodeWorker implements IWorker {
+
+	private readonly _worker: Worker;
+
+	constructor(callback: IWorkerCallback, onErrorCallback: (err: any) => void) {
+		this._worker = new Worker(
+			FileAccess.asFileUri('vs/platform/extensions/node/extensionHostStarterWorkerMain.js', require).fsPath,
+		);
+		this._worker.on('message', callback);
+		this._worker.on('error', onErrorCallback);
+		// this._worker.on('exit', (code) => {
+		// 	console.log(`worker exited with code `, code);
+		// });
+	}
+
+	getId(): number {
+		return 1;
+	}
+
+	postMessage(message: any, transfer: ArrayBuffer[]): void {
+		this._worker.postMessage(message, transfer);
+	}
+
+	dispose(): void {
+		this._worker.terminate();
+	}
+}
+
+class ExtensionHostStarterWorkerHost implements IExtensionHostStarterWorkerHost {
+	constructor(
+		@ILogService private readonly _logService: ILogService
+	) { }
+
+	public async logInfo(message: string): Promise<void> {
+		this._logService.info(message);
+	}
+}
+
+export class WorkerMainProcessExtensionHostStarter implements IDisposable, IExtensionHostStarter {
+	_serviceBrand: undefined;
+
+	private _proxy: ExtensionHostStarter | null;
+	private readonly _worker: SimpleWorkerClient<ExtensionHostStarter, IExtensionHostStarterWorkerHost>;
+	private _shutdown = false;
+
+	constructor(
+		@ILogService private readonly _logService: ILogService,
+		@ILifecycleMainService lifecycleMainService: ILifecycleMainService
+	) {
+		this._proxy = null;
+
+		const workerFactory: IWorkerFactory = {
+			create: (moduleId: string, callback: IWorkerCallback, onErrorCallback: (err: any) => void): IWorker => {
+				const worker = new NodeWorker(callback, onErrorCallback);
+				worker.postMessage(moduleId, []);
+				return worker;
+			}
+		};
+		this._worker = new SimpleWorkerClient<ExtensionHostStarter, IExtensionHostStarterWorkerHost>(
+			workerFactory,
+			'vs/platform/extensions/node/extensionHostStarterWorker',
+			new ExtensionHostStarterWorkerHost(this._logService)
+		);
+		this._initialize();
+
+		// Abnormal shutdown: terminate extension hosts asap
+		lifecycleMainService.onWillKill(async () => {
+			this._shutdown = true;
+			if (this._proxy) {
+				this._proxy.killAllNow();
+			}
+		});
+
+		// Normal shutdown: gracefully await extension host shutdowns
+		lifecycleMainService.onWillShutdown((e) => {
+			this._shutdown = true;
+			if (this._proxy) {
+				e.join(this._proxy.waitForAllExit(6000));
```

Comment:
Just a nit, but we seem to be giving extensions `5s` for graceful shutdown from within the extension host but we seem to be waiting `6s` from the outside.

Reply:
Yes, I think this time needs to be >5s. If the main process receives a shutdown request via `onWillShutdown` and it begins killing renderer processes, IMHO it is possible that the extension host might still take some time >0 until it realizes that the renderer socket is closed and then begins the shutdown. We can maybe shorten this to 5.5s, but I would like to keep this >5s.

Reply:
Thanks for the explanation makes sense.